core/src/services/icloud/backend.rs

```diff
@@ -0,0 +1,261 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::raw::*;
+use crate::services::icloud::client::Client;
+use crate::services::icloud::error::parse_error;
+use crate::services::icloud::session::{Session, SessionData};
+use crate::*;
+use crate::{Builder, Capability, Error, ErrorKind, Scheme};
+use async_trait::async_trait;
+use http::StatusCode;
+use log::debug;
+use std::collections::HashMap;
+use std::fmt::{Debug, Formatter};
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+#[derive(Default)]
+pub struct iCloudBuilder {
+    pub root: Option<String>,
+    pub apple_id: Option<String>,
+    pub password: Option<String>,
+    pub trust_token: Option<String>,
+    pub ds_web_auth_token: Option<String>,
+
+    pub http_client: Option<HttpClient>,
+}
+
+impl Debug for iCloudBuilder {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut d = f.debug_struct("iCloudBuilder");
+        d.field("root", &self.root);
+        d.field("apple_id", &self.apple_id);
+        d.field("password", &self.password);
+        d.field("trust_token", &self.trust_token);
+        d.field("ds_web_auth_token", &self.ds_web_auth_token);
+
+        d.finish_non_exhaustive()
+    }
+}
+
+impl iCloudBuilder {
+    pub fn root(&mut self, root: &str) -> &mut Self {
+        self.root = Some(root.to_string());
+        self
+    }
+
+    pub fn apple_id(&mut self, apple_id: &str) -> &mut Self {
+        self.apple_id = if apple_id.is_empty() {
+            None
+        } else {
+            Some(apple_id.to_string())
+        };
+
+        self
+    }
+
+    pub fn password(&mut self, password: &str) -> &mut Self {
+        self.password = if password.is_empty() {
+            None
+        } else {
+            Some(password.to_string())
+        };
+
+        self
+    }
+
+    pub fn trust_token(&mut self, trust_token: &str) -> &mut Self {
```

Comment:
It's better to provide a doc on how to fetch `trust_token` and `ds_web_auth_token`.